google/cloud/sql/connector/connector.py

```diff
@@ -23,43 +23,118 @@
 
+    """
 
-def connect(
-    instance_connection_string: str,
-    driver: str,
-    ip_types: IPTypes = IPTypes.PUBLIC,
-    enable_iam_auth: bool = False,
-    **kwargs: Any
-) -> Any:
-    """Prepares and returns a database connection object and starts a
-    background thread to refresh the certificates and metadata.
+    def __init__(
+        self,
+        ip_type: IPTypes = IPTypes.PUBLIC,
+        enable_iam_auth: bool = False,
+        timeout: int = 30,
+    ) -> None:
+        self._loop: asyncio.AbstractEventLoop = asyncio.new_event_loop()
+        self._thread: Thread = Thread(target=self._loop.run_forever, daemon=True)
+        self._thread.start()
+        self._keys: concurrent.futures.Future = asyncio.run_coroutine_threadsafe(
+            generate_keys(), self._loop
+        )
+        self._instances: Dict[str, InstanceConnectionManager] = {}
+
+        # set default params for connections
+        self._timeout = timeout
+        self._enable_iam_auth = enable_iam_auth
+        self._ip_type = ip_type
+
+    def connect(
+        self, instance_connection_string: str, driver: str, **kwargs: Any
+    ) -> Any:
+        """Prepares and returns a database connection object and starts a
+        background thread to refresh the certificates and metadata.
+
+        :type instance_connection_string: str
+        :param instance_connection_string:
+            A string containing the GCP project name, region name, and instance
+            name separated by colons.
+
+            Example: example-proj:example-region-us6:example-instance
+
+        :type driver: str
+        :param: driver:
+            A string representing the driver to connect with. Supported drivers are
+            pymysql, pg8000, and pytds.
+
+        :param kwargs:
+            Pass in any driver-specific arguments needed to connect to the Cloud
+            SQL instance.
+
+        :rtype: Connection
+        :returns:
+            A DB-API connection to the specified Cloud SQL instance.
+        """
+
+        # Initiate event loop and run in background thread.
+        #
+        # Create an InstanceConnectionManager object from the connection string.
+        # The InstanceConnectionManager should verify arguments.
+        #
+        # Use the InstanceConnectionManager to establish an SSL Connection.
+        #
+        # Return a DBAPI connection
+
+        if instance_connection_string in self._instances:
+            icm = self._instances[instance_connection_string]
+        else:
+            enable_iam_auth = kwargs.pop("enable_iam_auth", self._enable_iam_auth)
+            icm = InstanceConnectionManager(
+                instance_connection_string,
+                driver,
+                self._keys,
+                self._loop,
+                enable_iam_auth,
+            )
+            self._instances[instance_connection_string] = icm
+
+        ip_type = kwargs.pop("ip_types", self._ip_type)
```

Comment:
@kurtisvg What would you recommend here? Since the current version of the Python connector takes in `ip_types` as the parameter. Should I add support for both by adding a `kwargs.pop("ip_type")` check as well? Then update the README etc to use `ip_type` as the golden standard. Leaving in `ip_types` so we are still backwards compatible and doesn't break anything for current users.

Reply:
That's fine. You may want to add a log warning that ip_types is deprecated and we may remove it in a future release.